morpheus/_lib/src/stages/filter_detection.cpp

```diff
@@ -33,126 +36,193 @@
-                    const auto &stride = probs.get_stride();
-
-                    CHECK(probs.rank() == 2)
-                        << "C++ impl of the FilterDetectionsStage currently only supports two dimensional arrays";
-
-                    const std::size_t num_rows    = shape[0];
-                    const std::size_t num_columns = shape[1];
-
-                    // A bit ugly, but we cant get access to the rmm::device_buffer here. So make a copy
-                    auto tmp_buffer = std::make_shared<rmm::device_buffer>(probs.count() * probs.dtype_size(),
-                                                                           rmm::cuda_stream_per_thread);
-
-                    MRC_CHECK_CUDA(
-                        cudaMemcpy(tmp_buffer->data(), probs.data(), tmp_buffer->size(), cudaMemcpyDeviceToDevice));
-
-                    // Depending on the input the stride is given in bytes or elements, convert to elements
-                    auto tensor_stride = TensorUtils::get_element_stride<TensorIndex, std::size_t>(stride);
-
-                    // Now call the threshold function
-                    auto thresh_bool_buffer =
-                        MatxUtil::threshold(DevMemInfo{probs.count(), probs.dtype().type_id(), tmp_buffer, 0},
-                                            num_rows,
-                                            num_columns,
-                                            tensor_stride,
-                                            m_threshold,
-                                            true);
-                    std::vector<uint8_t> host_bool_values(num_rows);
-
-                    // Copy bools back to host
-                    MRC_CHECK_CUDA(cudaMemcpy(host_bool_values.data(),
-                                              thresh_bool_buffer->data(),
-                                              thresh_bool_buffer->size(),
-                                              cudaMemcpyDeviceToHost));
-
-                    // Only used when m_copy is true
-                    std::vector<std::pair<std::size_t, std::size_t>> selected_ranges;
-                    std::size_t num_selected_rows = 0;
-
-                    // We are slicing by rows, using num_rows as our marker for undefined
-                    std::size_t slice_start = num_rows;
-                    for (std::size_t row = 0; row < num_rows; ++row)
+    return [this](rxcpp::observable<sink_type_t> input, rxcpp::subscriber<source_type_t> output) {
+        return input.subscribe(rxcpp::make_observer<sink_type_t>(
+            [this, &output](sink_type_t x) {
+                BufferInfo buffer_info;
+                if (m_operate_on == FilterSource::TENSOR || m_operate_on == FilterSource::AUTO)
+                {
+                    buffer_info = get_tensor_buffer_info(x, m_field_name);
+                }
+                else
+                {
+                    buffer_info = get_column_buffer_info(x, m_field_name);
+                }
```

Comment:
`m_operate_on` shouldnt change during operation. Instead of branching inside of the hot path, return different `subscribe_fn_t` depending on the value. Then you dont need to check this value for each piece of data since it will never change.